src/Bicep.LangServer/BicepCompilationManager.cs

```diff
@@ -87,31 +91,66 @@ public void RefreshCompilation(DocumentUri documentUri)
                 {
                     UpsertCompilationInternal(documentUri, null, sourceFile);
                 }
+
                 return;
             }
 
             // TODO: This may cause race condition if the user is modifying the file at the same time
             // need to make a shallow copy so it counts as a different file even though all the content is identical
             // this was the easiest way to force the compilation to be regenerated
-            var shallowCopy = new BicepFile(compilationContext.Compilation.SourceFileGrouping.EntryPoint);
+            var shallowCopy = compilationContext.Compilation.SourceFileGrouping.EntryPoint.ShallowClone();
             UpsertCompilationInternal(documentUri, null, shallowCopy);
         }
 
-        public void UpsertCompilation(DocumentUri documentUri, int? version, string fileContents, string? languageId = null, bool triggeredByFileOpenEvent = false)
+        public void OpenCompilation(DocumentUri documentUri, int? version, string fileContents, string languageId)
         {
+            if (LanguageConstants.IsBicepOrParamsLanguage(languageId))
+            {
+                this.fileLanguageTracker.NotifyFileOpen(documentUri, languageId);
+            }
+
```

Comment:
Are you sure this is really necessary? What problem does it solve? `UpsertCompilationInternal` is very careful to update the active contexts atomically - splitting into 2 steps feels like it's going to introduce a potential for race conditions.

Reply:
It's trying to solve the fatal exception problem (number 1) from this discussion on this PR:
<img width="804" alt="image" src="https://user-images.githubusercontent.com/22460039/195237990-2c2b106f-a7a8-4e7b-8726-d2805e9b4494.png">

Basically the issue is this. To upsert the compilation for an untitled file when you have two language IDs possible, you need to have the ability to get the language ID for the previous version of that file. (ID is only available on file open.) If the previous version caused a fatal exception, you can't get it anymore and basically are forced to fail.

The language ID should be generally stable from file open to close for the vast majority of files. The potential for race conditions already existed before this change because that code was not truly atomic anyway. It's going to be hard to determine how much worse the race conditions will be given that one of the two steps involves infrequent updates (language IDs being generally stable for file lifetime).

Technically only untitled files require tracking of language IDs. For non-untitled files, the language can be inferred from the file extension and `SourceFileGroupingBuilder` is entirely file extension based anyway, which is making me question the value of trying too hard to be language ID based.

To reduce the chance for race conditions, I'm tempted to only use this new class for untitled files and switch everything to how it was before (basically use lang ID if available otherwise use file extension). Let's talk at our 1:1 tomorrow in case you have a better idea.

Reply:
There's also the option of creating a either a failed compilation (+ failed semantic model) or just a failed `CompilationContext`, which could carry the language ID across failed document versions.